craft_parts/overlays/overlay_manager.py

```diff
@@ -0,0 +1,114 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2021 Canonical Ltd.
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License version 3 as published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Overlay mount operations and package installation helpers."""
+
+import logging
+from pathlib import Path
+from typing import List, Optional
+
+from craft_parts.infos import ProjectInfo
+from craft_parts.parts import Part
+
+from .overlay_fs import OverlayFS
+
+logger = logging.getLogger(__name__)
+
+
+class OverlayManager:
+    """Execution time overlay mounting and package installation.
+
+    :param project_info: The project information.
+    :param part_list: A list of all parts in the project.
+    :param base_layer_dir: The directory containing the overlay base, or None
+        if the project doesn't use overlay parameters.
+    """
+
+    def __init__(
+        self,
+        *,
+        project_info: ProjectInfo,
+        part_list: List[Part],
+        base_layer_dir: Optional[Path]
+    ):
+        self._project_info = project_info
+        self._part_list = part_list
+        self._layer_dirs = [p.part_layer_dir for p in part_list]
+        self._overlay_mount_dir = project_info.overlay_mount_dir
+        self._overlay_fs: Optional[OverlayFS] = None
+        self._base_layer_dir = base_layer_dir
+
+    def mount_layer(self, part: Part, *, pkg_cache: bool = False) -> None:
+        """Mount the overlay step layer stack up to the given part.
+
+        :param part: The part corresponding to the topmost layer to mount.
+        :param pkg cache: Whether the package cache layer is enabled.
+        """
+        if not self._base_layer_dir:
+            return
```

Comment:
Here (and similar below), what should happen if not _base_layer_dir? I don't like this function silently doing nothing in this case. 

Is this a valid real case? I see from the signature and tests that `_base_layer_dir` may be None. My point is if instantiating `OverlayManager` with `base_layer_dir=None` will make all these to not do anything, what's the point of instantiating it?

Reply:
This is a consistency check to make sure the base layer is defined when the overlayfs mount is requested (the requests happen during overlay package installation, overlay script execution, and building of parts with overlay visibility). The method should never be called without the base dir and I'll raise runtime errors instead if that happens.